server/resp/src/main/resources/resp.gr

```diff
@@ -0,0 +1,77 @@
+// RESP 3.x protocol
+namespace resp3x;
+
+// Target class
+class org.infinispan.server.resp.RespDecoder extends BaseRespDecoder;
+
+// static final field values are literals
+constants org.infinispan.server.resp.RespConstants;
+
+// methods simpleString, bulkString, byteArray, byte...
+intrinsics org.infinispan.server.resp.Intrinsics;
+
+import java.nio.charset.StandardCharsets;
+import java.lang.UnsupportedOperationException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.util.Util;
+import org.infinispan.server.resp.logging.Log;
+import org.infinispan.server.resp.Intrinsics.Resp2LongProcessor;
+import org.infinispan.util.concurrent.CompletionStages;
+
+init {
+   private final List<byte[]> reusedList = new ArrayList<>();
+
+   public RespDecoder(RespRequestHandler initialHandler) {
+      this.requestHandler = initialHandler;
+   }
+}
+
+beforedecode {
+   // We cannot read more than one command at a time
+   if (!ctx.channel().config().isAutoRead()) {
+      log.tracef("Auto read was disabled, not reading next bytes");
+      return;
+   } else {
+      log.tracef("Auto read was enabled, reading next bytes");
+   }
+}
+
+// this is the root
+root request
+   : singleByte { if (singleByte != ARRAY) throw new UnsupportedOperationException("Only array types are supported, received: " + singleByte); reusedList.clear(); }
+     number
+     singleByte
+     command
+     arguments
+     execute
+   ;
+
+command returns String switch singleByte
+   : { BULK_STRING }? bulkString[longProcessor] { bulkString }
+   | { SIMPLE_STRING }? simpleString { simpleString }
+   | { throw new UnsupportedOperationException("Type not supported: " + singleByte); }
+   ;
+number: readNumber[longProcessor] { readNumber -= 1; };
+array returns byte[] switch singleByte
+   : { BULK_STRING }? bulkArray[longProcessor] { bulkArray }
+   // TODO: optimize this to not convert from bytes to string and back
+   | { SIMPLE_STRING }? simpleString { simpleString.getBytes() }
+   | { throw new UnsupportedOperationException("Type not supported: " + singleByte); }
+   ;
+arguments returns List<byte[]>
+   : { readNumber >= 1 }? { arguments = reusedList; } #readNumber ( singleByte array { arguments.add(array); } )
```

Comment:
Upon further thinking about it, I think we can just add an if the readNumber is greater than some cut off. And if the number is larger than that number it would create a new ArrayList and leave the reusedList alone. This way it would never need to be reclaimed.

The question is what size is this cut off. Maybe just 16 or 32?

Reply:
I think I will just leave it at 16. That should cover a lot of commands and allows for MSET to write to 8 keys and MGET to retrieve 16 without allocating an array.